### PR TITLE
Add a copy_file() API

### PR DESCRIPTION
**Pull request by cgwalters**
This copies both regular files and symbolic links, and makes
use of reflinks if available.

Came up in https://github.com/coreos/rpm-ostree/pull/2362/files

### CODE REVIEW

File: tests/basic.rs [hunk 1]
```diff
@@ -207,3 +207,18 @@ fn rmrf() -> anyhow::Result<()> {
 
     Ok(())
 }
+
+#[test]
+fn copy_path() -> Result<()> {
```
Comment (lucab) [inline tests/basic.rs:223]: I think this test is missing the overwrite-if-existing case for both files and symlinks.

File: src/lib.rs [hunk 2]
```diff
@@ -273,6 +293,107 @@ impl OpenatDirExt for openat::Dir {
     }
 }
 
+fn map_nix_error(e: nix::Error) -> io::Error {
+    match e.as_errno() {
+        Some(os_err) => return io::Error::from_raw_os_error(os_err as i32),
+        _ => return io::Error::new(io::ErrorKind::Other, e),
```
Comment (lucab) [inline src/lib.rs]: Nit: The `return` here and in the line above can be omitted.

File: src/lib.rs [hunk 3]
```diff
@@ -273,6 +293,107 @@ impl OpenatDirExt for openat::Dir {
...
+
+fn copy_regfile_inner(
+    src: &File,
+    srcmeta: &openat::Metadata,
+    dest: &mut FileWriter,
+) -> io::Result<()> {
+    // Go directly to the raw file so we can reflink
+    let destf = dest.writer.get_mut();
+    let _ = src.copy_to(destf)?;
+    if srcmeta.stat().st_uid != nix::unistd::getuid().as_raw() {
```
Comment (lucab) [inline src/lib.rs]: The `getuid()` here surprised me, and I suspect this condition fails to account for uid-vs-gid mismatches.
I would have expected something like `src.uid != dest.uid || src.gid != dest.gid`.

File: src/lib.rs [hunk 4]
```diff
@@ -273,6 +293,107 @@ impl OpenatDirExt for openat::Dir {
...
+fn impl_copy_symlink<D: openat::AsPath>(
+    dir: &openat::Dir,
+    s: &CStr,
+    srcmeta: &openat::Metadata,
+    d: D,
+) -> io::Result<()> {
+    let parent = Path::new(OsStr::from_bytes(s.to_bytes())).parent();
+    let symlinktarget = dir.read_link(s)?;
+    let tmplink = temp_symlink(dir, parent, symlinktarget.as_path())?;
+    if srcmeta.stat().st_uid != nix::unistd::getuid().as_raw() {
```
Comment (lucab) [inline src/lib.rs]: Same comment as above.

**Comment by cgwalters** [comment]
OK for rpm-ostree all we need immediately is the basic "copy file fd-relative", with the bonus feature of using reflinks if available.  I reworked things here to (almost) match `std::fs::copy()`, with the exception that symlinks aren't followed by default because that's how `openat` behaves.

Notably:
 - Drop attempts to `chown`
 - Drop (nontrivial) code for copying symlinks

More broadly speaking as noted in our team chat I am increasingly thinking that `openat` has some design flaws and is under-maintained; we shouldn't be building on it, but should instead be investigating https://github.com/bytecodealliance/cap-std

So for now let's just do the simple thing.

**Review by lucab — APPROVED** [review]

**Pull request merged**